Source/WebCore/dom/SpatialImageControls.cpp

```diff
@@ -0,0 +1,219 @@
+#include <wtf/text/AtomString.h>
+
+namespace WebCore {
+namespace SpatialImageControls {
+
+#if ENABLE(SPATIAL_IMAGE_CONTROLS)
+
+static const AtomString& spatialImageControlsElementIdentifier()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("spatial-image-controls"_s);
+    return identifier;
+}
+
+static const AtomString& spatialImageControlsButtonIdentifier()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("spatial-image-controls-button"_s);
+    return identifier;
+}
+
+bool hasSpatialImageControls(const HTMLElement& element)
+{
+    RefPtr shadowRoot = element.shadowRoot();
+    if (!shadowRoot || shadowRoot->mode() != ShadowRootMode::UserAgent)
+        return false;
+
+    return shadowRoot->hasElementWithId(spatialImageControlsElementIdentifier());
+}
+
+static RefPtr<HTMLElement> spatialImageControlsHost(const Node& node)
+{
+    RefPtr host = dynamicDowncast<HTMLElement>(node.shadowHost());
+    if (!host)
+        return nullptr;
+
+    return hasSpatialImageControls(*host) ? host : nullptr;
+}
+
+bool isSpatialImageControlsButtonElement(const Element& element)
+{
+    return spatialImageControlsHost(element) && element.getIdAttribute() == spatialImageControlsButtonIdentifier();
+}
+
+bool shouldHaveSpatialControls(HTMLImageElement& element)
+{
+    if (!element.document().settings().spatialImageControlsEnabled())
+        return false;
+
+    bool hasSpatialcontrolsAttribute = element.hasAttributeWithoutSynchronization(HTMLNames::controlsAttr);
+
+    auto* cachedImage = element.cachedImage();
+    if (!cachedImage)
+        return false;
+
+    auto* image = cachedImage->image();
+    if (!image)
+        return false;
+
+    return hasSpatialcontrolsAttribute && image->isSpatial();
+}
+
+void addShadowRootForSpatialImageControls(HTMLImageElement& imageElement)
+{
+    if (shouldHaveSpatialControls(imageElement) && !hasSpatialImageControls(imageElement)) {
+        imageElement.protectedDocument()->checkedEventLoop()->queueTask(TaskSource::InternalAsyncTask, [weakElement = WeakPtr { imageElement }] {
+            RefPtr element = weakElement.get();
+            element->ensureUserAgentShadowRoot();
+        });
+    }
+}
+
+void ensureSpatialControls(HTMLImageElement& imageElement)
+{
+    if (!shouldHaveSpatialControls(imageElement))
+        return;
+
+    imageElement.protectedDocument()->checkedEventLoop()->queueTask(TaskSource::InternalAsyncTask, [weakElement = WeakPtr { imageElement }] {
+        RefPtr element = weakElement.get();
+        RefPtr shadowRoot = element->shadowRoot();
+        Ref document = element->document();
+
+        Ref controlLayer = HTMLDivElement::create(document.get());
+        controlLayer->setIdAttribute(spatialImageControlsElementIdentifier());
+        controlLayer->setAttributeWithoutSynchronization(HTMLNames::contenteditableAttr, falseAtom());
+        controlLayer->setAttributeWithoutSynchronization(HTMLNames::styleAttr, "position: relative;"_s);
```

Comment:
The extra container (with `position: relative`) is necessary for correct positioning of the button.

To include it in the UA stylesheet I add I'd have to move the style block out of being within the `div` so it could apply to the `div`, and I ran into issues when removing the controls (when the `controls` attribute gets removed) when the shadow root had two immediate children (`style` and `div`). I could keep looking for a workaround if this is bad form though?

Reply:
nah, that makes sense!